### PR TITLE
SM-5453: SNS Service - send message to topic

### PR DESCRIPTION
**Pull request by Kellym-Kainos**
## Description
Extend skeleton Lambda to send SNS messages

- Added SNS service to Lambda
- Send simple message to specified Work Start SNS Topic 

### Related PRs

**Terraform**: https://github.com/departmentfortransport/street-manager-terraform/pull/1204 
[Error with DLQ setup for SNS]

## Coding Standards

Style guide: https://github.com/departmentfortransport/street-manager/wiki/Street-Manager-Style-Guide
Coding best practices: https://github.com/departmentfortransport/street-manager/wiki/Street-Manager-Coding-Best-Practices

## Checklist

- [x] All unit tests passing
- [x] Code coverage suitable
- [x] Branch named {feature|hotfix|task}/{SM-.*}
- [x] If your pull request depends on any other, please link them
- [ ] Changes approved by your team
- [ ] Changes approved by another team
- [ ] Significant events are being logged e.g. error scenarios
- [ ] Commit messages are meaningful
- [ ] Target is green https://circleci.com/gh/departmentfortransport/street-manager-event-notifier/tree/master
- [ ] Int UI tests passing https://circleci.com/gh/departmentfortransport/street-manager-int/tree/master
- [ ] Add `DO NOT MERGE` if you want to postpone merge

### CODE REVIEW

File: src/app.ts [hunk 1]
```diff
@@ -3,20 +3,43 @@ import iocContainer from './ioc'
...
 import Logger from './utils/logger'
+import SNSService from './services/aws/snsService'
+import { SQS } from 'aws-sdk'
 
 let knex: Knex
 const logger: Logger = iocContainer.get<Logger>(TYPES.Logger)
 
 async function main(event) {
-  logger.log(`Recieved message: ${JSON.stringify(event)}`)
+    const sqsMessage: SQS.Types.Message = <SQS.Types.Message>event
```
Comment (cpaul92) [inline src/app.ts:13]: indents out here

File: tests/unit/services/aws/snsServiceTests.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
+import 'mocha'
+import * as sinon from 'sinon'
+import SNSService from '../../../../src/services/aws/snsService'
+import { SNS } from 'aws-sdk'
+import { assert } from 'chai'
+import Logger from '../../../../src/utils/logger'
+import { mock} from 'ts-mockito'
```
Comment (cpaul92) [inline tests/unit/services/aws/snsServiceTests.ts:7]:
```suggestion
import { mock } from 'ts-mockito'
```

File: src/services/aws/snsService.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    const params: SNS.PublishInput = {
+      Message: message,
+      TopicArn: topic
+    }
+
+    try {
+      await super.toAWSPromise<SNS.PublishInput, SNS.PublishResponse>(this.sns.publish, params)
+    } catch (err) {
+      this.logger.error(err)
+      throw new Error(err)
```
Comment (matthewdoreilly) [inline src/services/aws/snsService.ts:25]: `return Promise.reject(err)`?

File: tests/unit/services/aws/snsServiceTests.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+      const message = JSON.stringify({test: 'message'})
+      const arn = 'topicArn'
+
+      await service.publishMessage(message, arn)
+
+      const input: SNS.PublishInput = publish.lastCall.args[0]
+
+      assert.equal(input.Message, message)
+      assert.equal(input.TopicArn, arn)
+    })
```
Comment (matthewdoreilly) [inline tests/unit/services/aws/snsServiceTests.ts:43]: Can we add a test for when sending the message to SNS fails?

File: src/services/aws/snsService.ts (hunk 3 above)
Comment (Kellym-Kainos) [reply src/services/aws/snsService.ts:25]: hm yeah can return one here and throw at the app.ts level

File: README.md [hunk 5]
```diff
@@ -10,9 +10,12 @@ Lambda function that handles the sending of notificiations from Street Manager v
 
 ### 2. Environment variables
```
[the remark this answers is not in the document]
Comment (Kellym-Kainos) [reply README.md:11]: yep can do

File: src/app.ts [hunk 6]
```diff
@@ -3,20 +3,41 @@ import iocContainer from './ioc'
...
+import SNSService from './services/aws/snsService'
+import { SQS } from 'aws-sdk'
 
-let knex: Knex
 const logger: Logger = iocContainer.get<Logger>(TYPES.Logger)
 
 async function main(event) {
-  logger.log(`Recieved message: ${JSON.stringify(event)}`)
+  const sqsMessage: SQS.Types.Message = <SQS.Types.Message>event
+  logger.log(`Recieved message: ${JSON.stringify(sqsMessage)}`)
```
Comment (matthewdoreilly) [inline src/app.ts:13]:
```suggestion
  logger.log(`Received message: ${JSON.stringify(sqsMessage)}`)
```

Comment (Kellym-Kainos) [reply src/app.ts:13]: i before e except after c 😏

**Pull request merged**